### PR TITLE
Architect browser schema

### PR DESCRIPTION
**Pull request by hansl**
refactor(@angular-devkit/build-angular): clean up some interfaces and schemas

Remove the manually maintained schema.d.ts from the browser builder, and use
the Schema JSON file to generate it. This had a lot of repercussions around
the whole build-angular code base and the different interfaces that were
manually kept.

### CODE REVIEW

File: packages/angular_devkit/build_angular/test/browser/source-map_spec_large.ts [hunk 1]
```diff
@@ -120,8 +121,10 @@ describe('Browser Builder source map', () => {
   it('supports styles only sourcemaps', (done) => {
     const overrides: Partial<BrowserBuilderSchema> = {
       sourceMap: {
```
Comment (alan-agius4) [inline packages/angular_devkit/build_angular/test/browser/source-map_spec_large.ts]: why are these options required now?

File: packages/angular_devkit/build_angular/src/angular-cli-files/models/webpack-configs/browser.ts [hunk 2]
```diff
@@ -67,9 +67,9 @@ export function getBrowserConfig(wco: WebpackConfigOptions) {
 
   if (!isEval && (scriptsSourceMap || stylesSourceMap)) {
     extraPlugins.push(getSourceMapDevTool(
-      scriptsSourceMap,
-      stylesSourceMap,
-      hiddenSourceMap,
+      scriptsSourceMap === undefined ? true : scriptsSourceMap,
```
Comment (alan-agius4) [inline packages/angular_devkit/build_angular/src/angular-cli-files/models/webpack-configs/browser.ts]: Can we use the normalizeSourceMapOptions method for this cases, setting these default in multiple places seems risky.

File: packages/angular_devkit/build_angular/src/angular-cli-files/models/webpack-configs/server.ts [hunk 3]
```diff
@@ -21,9 +21,9 @@ export function getServerConfig(wco: WebpackConfigOptions) {
     const { scripts, styles, hidden } = wco.buildOptions.sourceMap;
 
     extraPlugins.push(getSourceMapDevTool(
-      scripts,
-      styles,
-      hidden,
+      scripts === undefined ? true : scripts,
+      styles === undefined ? true : styles,
+      hidden === undefined ? true : hidden,
```
Comment (alan-agius4) [inline packages/angular_devkit/build_angular/src/angular-cli-files/models/webpack-configs/server.ts]: When undefined` it's `true`? this is changing the behaviour

File: packages/angular_devkit/build_angular/src/dev-server/index.ts [hunk 4]
```diff
@@ -105,7 +86,7 @@ export class DevServerBuilder implements Builder<DevServerBuilderOptions> {
         const serverAddress = url.format({
           protocol: options.ssl ? 'https' : 'http',
           hostname: options.host === '0.0.0.0' ? 'localhost' : options.host,
-          port: options.port.toString(),
+          port: (options.port || 4200).toString(),
```
Comment (alan-agius4) [inline packages/angular_devkit/build_angular/src/dev-server/index.ts]: Why not use the defaults in the schema?

File: packages/angular_devkit/build_angular/src/angular-cli-files/models/webpack-configs/browser.ts (hunk 2 above)
Comment (hansl) [reply packages/angular_devkit/build_angular/src/angular-cli-files/models/webpack-configs/browser.ts]: We should, or at least add MORE normalizing methods.

File: packages/angular_devkit/build_angular/src/angular-cli-files/models/webpack-configs/server.ts (hunk 3 above)
Comment (hansl) [reply packages/angular_devkit/build_angular/src/angular-cli-files/models/webpack-configs/server.ts]: Current behavior is to default to true, no? The schema says it's true by default, the type says it's optional.

File: packages/angular_devkit/build_angular/src/dev-server/index.ts (hunk 4 above)
Comment (hansl) [reply packages/angular_devkit/build_angular/src/dev-server/index.ts]: The type is optional, so typescript complains here. I need to figure out how to better do this.

File: packages/angular_devkit/build_angular/test/browser/source-map_spec_large.ts (hunk 1 above)
Comment (hansl) [reply packages/angular_devkit/build_angular/test/browser/source-map_spec_large.ts]: They're not, I just went back and forth between required and not required.

File: packages/angular_devkit/build_angular/src/angular-cli-files/models/webpack-configs/server.ts (hunk 3 above)
Comment (alan-agius4) [reply packages/angular_devkit/build_angular/src/angular-cli-files/models/webpack-configs/server.ts]: It’s true for souceMap but not when it’s a complex object, when it is a complex object at the moment, hidden is also false.

**Pull request merged**